### ISSUE TITLE
abort when using gstreamer

### ISSUE THREAD

**Issue by Michael65589** [labwc/labwc#1522]
I'm using Labwc in master branch with wlroots v0.17
When I'm trying to use gstreamer, labwc breaks. 
GStreamer pipeline: gst-launch-1.0 videotestsrc ! waylandsink

[1034691.621]  -> xdg_toplevel@19.configure(0, 0, array[0])
[1034691.638]  -> xdg_surface@18.configure(2)
[1034692.392] xdg_wm_base@9.pong(3)
[1034692.418] xdg_surface@18.ack_configure(2)
[1034692.434] wl_surface@3.commit()
[1034692.468] zwp_linux_dmabuf_v1@5.create_params(new id zwp_linux_buffer_params_v1@17)
[1035036.913] zwp_linux_buffer_params_v1@17.add(fd 36, 0, 0, 1280, 0, 0)
[1035036.956] zwp_linux_buffer_params_v1@17.add_dtrc_meta(0, 0)
00:00:15.480 [wayland] listener function for opcode 4 of zwp_linux_buffer_params_v1 is NULL
Aborted

**Comment by Consolatis** [comment]
Could you do a grep `zwp_linux` on a `WAYLAND_DEBUG=1` log?

This seems like gstreamer using an unsupported version of the protocol. Either because labwc or wlroots for some reason announce a newer version of the protocol that they support or because gstreamer simply doesn't care (in that case there is still a bug in labwc or wlroots because we should not abort in that case but just kill the client).

**Comment by Michael65589** [comment]
Ok here is the result:

$cat labwc.log | grep zwp_linux
[1021274.205]  -> wl_registry@2.global(3, "zwp_linux_dmabuf_v1", 4)
[1021275.229] wl_registry@2.bind(3, "zwp_linux_dmabuf_v1", 3, new id [unknown]@5)
[1021275.246]  -> zwp_linux_dmabuf_v1@5.modifier(538982482, 0, 0)
[1021275.261]  -> zwp_linux_dmabuf_v1@5.modifier(538982482, 100663296, 4)
[1021275.275]  -> zwp_linux_dmabuf_v1@5.modifier(538982482, 16777215, 4294967295)
[1021275.291]  -> zwp_linux_dmabuf_v1@5.modifier(943215175, 0, 0)
[1021275.306]  -> zwp_linux_dmabuf_v1@5.modifier(943215175, 100663296, 4)
[1021275.320]  -> zwp_linux_dmabuf_v1@5.modifier(943215175, 16777215, 4294967295)
[1021275.335]  -> zwp_linux_dmabuf_v1@5.modifier(842093144, 0, 0)
[1021275.348]  -> zwp_linux_dmabuf_v1@5.modifier(842093144, 100663296, 4)
[1021275.364]  -> zwp_linux_dmabuf_v1@5.modifier(842093144, 16777215, 4294967295)
[1021275.378]  -> zwp_linux_dmabuf_v1@5.modifier(842089048, 0, 0)
[1021275.394]  -> zwp_linux_dmabuf_v1@5.modifier(842089048, 100663296, 4)
[1021275.409]  -> zwp_linux_dmabuf_v1@5.modifier(842089048, 16777215, 4294967295)
[1021275.425]  -> zwp_linux_dmabuf_v1@5.modifier(842093121, 0, 0)
[1021275.439]  -> zwp_linux_dmabuf_v1@5.modifier(842093121, 100663296, 4)
[1021275.454]  -> zwp_linux_dmabuf_v1@5.modifier(842093121, 16777215, 4294967295)
[1021275.468]  -> zwp_linux_dmabuf_v1@5.modifier(842089025, 0, 0)
[1021275.481]  -> zwp_linux_dmabuf_v1@5.modifier(842089025, 100663296, 4)
[1021275.495]  -> zwp_linux_dmabuf_v1@5.modifier(842089025, 16777215, 4294967295)
[1021275.510]  -> zwp_linux_dmabuf_v1@5.modifier(892424792, 0, 0)
[1021275.525]  -> zwp_linux_dmabuf_v1@5.modifier(892424792, 100663296, 4)
[1021275.538]  -> zwp_linux_dmabuf_v1@5.modifier(892424792, 16777215, 4294967295)
[1021275.554]  -> zwp_linux_dmabuf_v1@5.modifier(892420696, 0, 0)
[1021275.568]  -> zwp_linux_dmabuf_v1@5.modifier(892420696, 100663296, 4)
[1021275.581]  -> zwp_linux_dmabuf_v1@5.modifier(892420696, 16777215, 4294967295)
[1021275.595]  -> zwp_linux_dmabuf_v1@5.modifier(892424769, 0, 0)
[1021275.645]  -> zwp_linux_dmabuf_v1@5.modifier(892424769, 100663296, 4)
[1021275.660]  -> zwp_linux_dmabuf_v1@5.modifier(892424769, 16777215, 4294967295)
[1021275.674]  -> zwp_linux_dmabuf_v1@5.modifier(892420673, 0, 0)
[1021275.689]  -> zwp_linux_dmabuf_v1@5.modifier(892420673, 100663296, 4)
[1021275.703]  -> zwp_linux_dmabuf_v1@5.modifier(892420673, 16777215, 4294967295)
[1021275.719]  -> zwp_linux_dmabuf_v1@5.modifier(909199186, 0, 0)
[1021275.732]  -> zwp_linux_dmabuf_v1@5.modifier(909199186, 100663296, 4)
[1021275.747]  -> zwp_linux_dmabuf_v1@5.modifier(909199186, 16777215, 4294967295)
[1021275.761]  -> zwp_linux_dmabuf_v1@5.modifier(909199170, 0, 0)
[1021275.775]  -> zwp_linux_dmabuf_v1@5.modifier(909199170, 100663296, 4)
[1021275.788]  -> zwp_linux_dmabuf_v1@5.modifier(909199170, 16777215, 4294967295)
[1021275.802]  -> zwp_linux_dmabuf_v1@5.modifier(875713112, 0, 0)
[1021275.817]  -> zwp_linux_dmabuf_v1@5.modifier(875713112, 100663296, 4)
[1021275.831]  -> zwp_linux_dmabuf_v1@5.modifier(875713112, 16777215, 4294967295)
[1021275.849]  -> zwp_linux_dmabuf_v1@5.modifier(875709016, 0, 0)
[1021275.862]  -> zwp_linux_dmabuf_v1@5.modifier(875709016, 100663296, 4)
[1021275.878]  -> zwp_linux_dmabuf_v1@5.modifier(875709016, 16777215, 4294967295)
[1021275.891]  -> zwp_linux_dmabuf_v1@5.modifier(875714642, 0, 0)
[1021275.905]  -> zwp_linux_dmabuf_v1@5.modifier(875714642, 100663296, 4)
[1021275.919]  -> zwp_linux_dmabuf_v1@5.modifier(875714642, 16777215, 4294967295)
[1021275.934]  -> zwp_linux_dmabuf_v1@5.modifier(875714626, 0, 0)
[1021275.948]  -> zwp_linux_dmabuf_v1@5.modifier(875714626, 100663296, 4)
[1021275.961]  -> zwp_linux_dmabuf_v1@5.modifier(875714626, 16777215, 4294967295)
[1021275.977]  -> zwp_linux_dmabuf_v1@5.modifier(875713089, 0, 0)
[1021275.990]  -> zwp_linux_dmabuf_v1@5.modifier(875713089, 100663296, 4)
[1021276.004]  -> zwp_linux_dmabuf_v1@5.modifier(875713089, 16777215, 4294967295)
[1021276.018]  -> zwp_linux_dmabuf_v1@5.modifier(875708993, 0, 0)
[1021276.033]  -> zwp_linux_dmabuf_v1@5.modifier(875708993, 100663296, 4)
[1021276.084]  -> zwp_linux_dmabuf_v1@5.modifier(875708993, 16777215, 4294967295)
[1021276.099]  -> zwp_linux_dmabuf_v1@5.modifier(875708754, 0, 0)
[1021276.115]  -> zwp_linux_dmabuf_v1@5.modifier(875708754, 100663296, 4)
[1021276.128]  -> zwp_linux_dmabuf_v1@5.modifier(875708754, 16777215, 4294967295)
[1021276.144]  -> zwp_linux_dmabuf_v1@5.modifier(875708738, 0, 0)
[1021276.158]  -> zwp_linux_dmabuf_v1@5.modifier(875708738, 100663296, 4)
[1021276.173]  -> zwp_linux_dmabuf_v1@5.modifier(875708738, 16777215, 4294967295)
[1021276.188]  -> zwp_linux_dmabuf_v1@5.modifier(1448695129, 16777215, 4294967295)
[1021276.202]  -> zwp_linux_dmabuf_v1@5.modifier(1431918169, 16777215, 4294967295)
[1021276.216]  -> zwp_linux_dmabuf_v1@5.modifier(1498831189, 16777215, 4294967295)
[1021276.230]  -> zwp_linux_dmabuf_v1@5.modifier(1498765654, 16777215, 4294967295)
[1021276.245]  -> zwp_linux_dmabuf_v1@5.modifier(1448433985, 16777215, 4294967295)
[1021276.259]  -> zwp_linux_dmabuf_v1@5.modifier(842094158, 16777215, 4294967295)
[1021276.277]  -> zwp_linux_dmabuf_v1@5.modifier(825382478, 16777215, 4294967295)
[1021276.291]  -> zwp_linux_dmabuf_v1@5.modifier(909203022, 16777215, 4294967295)
[1021276.307]  -> zwp_linux_dmabuf_v1@5.modifier(825644622, 16777215, 4294967295)
[1021276.321]  -> zwp_linux_dmabuf_v1@5.modifier(842093913, 16777215, 4294967295)
[1021276.335]  -> zwp_linux_dmabuf_v1@5.modifier(842094169, 16777215, 4294967295)
[1021276.349]  -> zwp_linux_dmabuf_v1@5.modifier(538982476, 0, 0)
[1021276.364]  -> zwp_linux_dmabuf_v1@5.modifier(538982476, 100663296, 4)
[1021276.378]  -> zwp_linux_dmabuf_v1@5.modifier(538982476, 16777215, 4294967295)
[1021276.392]  -> zwp_linux_dmabuf_v1@5.modifier(943210828, 0, 0)
[1021276.406]  -> zwp_linux_dmabuf_v1@5.modifier(943210828, 100663296, 4)
[1021276.420]  -> zwp_linux_dmabuf_v1@5.modifier(943210828, 16777215, 4294967295)
[1021288.644] zwp_linux_dmabuf_v1@5.create_params(new id zwp_linux_buffer_params_v1@17)
[1021288.662] zwp_linux_buffer_params_v1@17.add(fd 37, 0, 0, 1280, 0, 0)
[1021288.680] zwp_linux_buffer_params_v1@17.add_dtrc_meta(0, 0)
00:00:09.644 [INFO] [wayland] listener function for opcode 4 of zwp_linux_buffer_params_v1 is NULL

**Comment by Consolatis** [comment]
> `[1021288.680] zwp_linux_buffer_params_v1@17.add_dtrc_meta(0, 0)`

This doesn't seem to be part of the spec: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/blob/main/unstable/linux-dmabuf/linux-dmabuf-unstable-v1.xml?ref_type=heads#L193

Not sure where that is coming from.
Just to ensure we are on the same page: it is labwc that is aborting here, no the gstreamer pipeline?

**Comment by Michael65589** [comment]
I started labwc in a console and gstreamer in another. When I enter the pipeline command I see that labwc aborts.

Console output labwc:
00:00:14.106 [wayland] listener function for opcode 4 of zwp_linux_buffer_params_v1 is NULL
Aborted

Console output pipeline
~$ gst-launch-1.0 -v videotestsrc ! waylandsink display=wayland-0
Setting pipeline to PAUSED ...
Pipeline is PREROLLING ...
/GstPipeline:pipeline0/GstVideoTestSrc:videotestsrc0.GstPad:src: caps = video/x-raw, format=(string)BGRA, width=(int)320, height=(int)240, framerate=(fraction)30/1, multiview-mode=(string)mono, pixel-aspect-ratio=(fraction)1/1, interlace-mode=(string)progressive
/GstPipeline:pipeline0/GstWaylandSink:waylandsink0.GstPad:sink: caps = video/x-raw, format=(string)BGRA, width=(int)320, height=(int)240, framerate=(fraction)30/1, multiview-mode=(string)mono, pixel-aspect-ratio=(fraction)1/1, interlace-mode=(string)progressive
Pipeline is PREROLLED ...
Setting pipeline to PLAYING ...
Redistribute latency...
New clock: GstSystemClock
0:00:00.0 / 99:99:99.

**Comment by Consolatis** [comment]
Could you post a stack trace for labwc?
Also: what distro and gstreamer version is this happening with?

**Comment by Michael65589** [comment]
How can I create a stack trace? (sorry for asking)

I am working on a Yocto based Image. 
Linux-Kernel 5.15.52
Wlroots 0.17
Wayland-Protocols 1.32
Wayland 1.22.0
GStreamer version 1.20.3

**Comment by Consolatis** [comment]
Stack trace is not required anymore. It seems like libwayland is causing the abort due to random changes to the protocol XML.

Something is very wrong with the build environment for gstreamer and likely also for wlroots. Do you have a link for the sources used for these two packages (and also for the wayland-protocols package)?

**Comment by Michael65589** [comment]
GStreamer: 
https://gstreamer.freedesktop.org/src/gstreamer/gstreamer-1.20.3.tar.xz
wlroots:
SRC_URI = "git://gitlab.freedesktop.org/wlroots/wlroots.git;branch=master;protocol=https"
SRCREV = "767eedd3cbe9900687bf3b82236320dcd7b77aae"
PV = "0.17.0"
(https://github.com/MarkusVolk/meta-wayland/blob/master/recipes-wlroots/wlroots/wlroots-0.17.bb)
wayland-protocols:
git://github.com/nxp-imx/wayland-protocols-imx.git;protocol=https;branch=wayland-protocols-imx-1.32

**Comment by Consolatis** [comment]
Your wayland-protocols package is broken. It contains changes that are neither official nor even properly versioned. You can't modify an official wayland protocol in this way without renaming it. Please use a proper upstream wayland-protocols package.

**Comment by Michael65589** [comment]
The main problem is that changes itself are wrong or is the kind of implementation wrong? I do not exaclty know what I should write to the NXP employee to get help.

**Comment by Consolatis** [comment]
It seems NXP only added two new protocols (which is fine) and changed the dmabuf one (which isn't fine).
As neither wlroots nor labwc is using any of the 2 new protocols, you should be fine using the upstream wayland-protocols package.

**Comment by Michael65589** [comment]
This one you mean : https://gitlab.freedesktop.org/wayland/wayland-protocols
Right?

**Comment by Consolatis** [comment]
> This one you mean : https://gitlab.freedesktop.org/wayland/wayland-protocols Right?

Correct.

**Comment by Consolatis** [comment]
> In my opinion they did the changes to have access the dma and gpu system of the i.MX8.

Yes, that is likely. However they did it at the wrong place. They should first and foremost work [upstream](https://gitlab.freedesktop.org/wayland/wayland-protocols) so everybody in the wayland ecosystem benefits from the changes and even if they decide to go their own ways they should create a new protocol for that rather than modifying an existing one.

> If they can't or won't help me is there a possibility to compile wlroots and labwc against the new protocol of NXP?

Their changes will not have any effect when it is not also implemented in the compositor and client applications. So compiling wlroots against the new protocol would not help at all without also implementing whatever their addition does in wlroots and all client applications that should make use of it. Did you try to just use the upstream wayland-protocols for your yocto build? That likely works fine.

**Comment by Michael65589** [comment]
Yes I tried the upstream wayland-protocols. But I have to change the Gstreamer-version and Weston compositor as well. NXP did changes in Gstreamer and the plugins as in Weston.
Now I can run labwc with, for example, an Qt application but when I start a gstreamer pipeline I am not able to set the size and position of the window. It always appears in the center of the screen. The waylandsink hasn't any properties to set x,y and with, height.
In addition I tried to move the orientation of the sreen from landscape to portrait. I added the tablet item to the rc.xml and set rotate to 90 but without success. Was this the right way, or is there another way to rotate the screen?

**Comment by Consolatis** [comment]
> Was this the right way, or is there another way to rotate the screen?

`wlr-randr --output HDMI-A-1 --transform 90`

> Now I can run labwc with, for example, an Qt application but when I start a gstreamer pipeline I am not able to set the size and position of the window. It always appears in the center of the screen. The waylandsink hasn't any properties to set x,y and with, height.

You could add a [window rule](https://labwc.github.io/labwc-config.5.html#window_rules) for that and then use the [MoveTo](https://labwc.github.io/labwc-actions.5.html#entry_action_name=moveto_x=_y=) action.

**Comment by Michael65589** [comment]
Hi,
how do I get the app_id or identifier of a window? I tried to start labwc with parameter -d but I can't find any hint in the output when I start an application.
My objective is to move a window of an application to a fixed position

**Comment by stefonarch** [comment]
https://sr.ht/~leon_plickat/lswt/ is nice for that.

**Comment by Michael65589** [comment]
WAYLAND_DISPLAY=wayland-0 ./lswt -j
[
    {
        "title": "<NULL>",
        "app_id": "<NULL>",
        "maximized": false,
        "minimized": false,
        "activated": true,
        "fullscreen": false,
        "outputs": [ "LVDS-1" ]
    }
]

Is there a implementation error of wayland on gstreamer side? Why is "app_id": "<NULL>" ??

**Comment by Consolatis** [comment]
> A last question to the experts. In general wayland does not support the possibility to have a start position of the window? It is always the wayland-client (compositor) task to handle that?

The design of wayland puts the compositor in full control of the window positions. The wayland clients (e.g. gstreamer) are not even aware of their own position.

> It would be much more nicer when the application, in my case the GStreamer plugin can decide where the window appears. This have much more flexibility.
> When always the compositor decide where the window appears and I only can control that by a configure file, I only have one possible position.

That is not possible with wayland, at least with the official wayland protocols.
A (hacky) workaround could be to use different app_id's and restart gstreamer (I don't think labwc re-applies window rules on app_id changes during runtime but that could potentially be implemented) or you could try to run your gstreamer pipeline via xwayland instead (e.g. using a X11 sink) but I don't know how well that would work.

Another theoretical option could be to have your own fullscreen application which includes the gstreamer output as a sub-surface, in that case you'd have full control over its position. I lack any experience in that area though so I am not even sure if its possible and what the performance impact of that would be.
